R/countdown.R

```diff
@@ -300,33 +306,72 @@ countdown_fullscreen <- function(
   )
 }
 
+# ---- Style Countdown ----
+
+#' @describeIn countdown Set global default countdown timer styles using CSS.
+#'   Use this function to globally style all countdown timers in a document or
+#'   app. Individual timers can still be customized.
+#' @param .selector In `style_countdown()`: the CSS selector to which the styles
+#'   should be applied. The default is `:root` for global styles, but you can
+#'   also provide a custom class name to create styles for a particular class.
+#' @export
+style_countdown <- function(
+  font_size = "3rem",
+  margin = "0.6em",
+  padding = "10px 15px",
+  box_shadow = "0px 4px 10px 0px rgba(50, 50, 50, 0.4)",
+  border_width = "0.1875 rem",
+  border_radius = "0.9rem",
+  line_height = "1",
+  color_border = "#ddd",
+  color_background = "inherit",
+  color_text = "inherit",
+  color_running_background = "#43AC6A",
+  color_running_border = prismatic::clr_darken(color_running_background, 0.1),
+  color_running_text = NULL,
+  color_finished_background = "#F04124",
+  color_finished_border = prismatic::clr_darken(color_finished_background, 0.1),
+  color_finished_text = NULL,
+  color_warning_background = "#E6C229",
+  color_warning_border = prismatic::clr_darken(color_warning_background, 0.1),
+  color_warning_text = NULL,
+  .selector = "root"
+) {
+  # get user args and defaults of current call
+  arg_names <- names(formals(style_countdown))
+  arg_names <- setdiff(arg_names, ".selector")
+  dots <- lapply(arg_names, get, envir = environment())
+  names(dots) <- arg_names
+
+  css_vars <- make_countdown_css_vars(.list = dots)
+  declarations <- css(!!!css_vars)
 
-make_unique_id <- function() {
-  with_private_seed <- utils::getFromNamespace("withPrivateSeed", "htmltools")
-  with_private_seed({
-    rand_id <- as.hexmode(sample(256, 4, replace = TRUE) - 1)
-    paste(format(rand_id, width=2), collapse = "")
-  })
+  tags$style(HTML(sprintf(":%s {%s}", .selector, declarations)))
 }
 
-validate_html_id <- function(id) {
-  stop_because <- function(...) {
-    stop(paste0('"', id, '" is not a valid HTML ID: ', ...))
-  }
-  if (!grepl("^[a-zA-Z]", id)) {
-    stop_because("Must start with a letter")
+make_countdown_css_vars <- function(..., .list = list()) {
```

Comment:
This is a _really_ elegant and concise way of setting the CSS variables.

Reply:
Thanks! Sometimes it pays to be lazy (and to be consistent in naming things) 😉